kernel/aster-nix/src/util/net/socket.rs

```diff
@@ -68,3 +69,33 @@ bitflags! {
         const SOCK_CLOEXEC = 1 << 19;
     }
 }
+
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Pod)]
+pub struct CUserMsgHdr {
+    /// Pointer to socket address structure
+    pub msg_name: Vaddr,
+    /// Size of socket address
+    pub msg_namelen: i32,
+    /// Scatter/Gather iov array
+    pub msg_iov: Vaddr,
+    /// The # of elements in msg_iov
+    pub msg_iovlen: u32,
+    /// Ancillary data
+    pub msg_control: Vaddr,
+    /// Ancillary data buffer length
+    pub msg_controllen: u32,
+    /// Flags on received message
+    pub msg_flags: u32,
+}
+
+impl CUserMsgHdr {
+    pub fn read_socket_addr(&self) -> Result<Option<SocketAddr>> {
+        if self.msg_name == 0 {
+            return Ok(None);
+        }
+
+        let socket_addr = read_socket_addr_from_user(self.msg_name, self.msg_namelen as usize)?;
+        Ok(Some(socket_addr))
+    }
```

Comment:
Perhaps we have a method that creates the `IoVecIter` here? We invoke the long expression `IoVecIter::new(c_user_msghdr.msg_iov, c_user_msghdr.msg_iovlen as usize)` multiple times.